File: go-service/internal/infrastructure/repository/product_repository.go
package repository

import (
	"context"

	"github.com/example/interview-stack/go-service/internal/domain"
)

// ProductRepository bridges the domain port with the configured store implementation.
type ProductRepository struct {
	store ProductStore
	clock domain.Clock
}

func NewProductRepository(store ProductStore, clock domain.Clock) *ProductRepository {
	return &ProductRepository{
		store: store,
		clock: clock,
	}
}

func (r *ProductRepository) ListProducts(ctx context.Context) ([]domain.Product, error) {
	products, err := r.store.ListProducts(ctx)
	if err != nil {
		return nil, domain.NewInternalError("failed to list products", err)
	}
	return products, nil
}

func (r *ProductRepository) GetProduct(ctx context.Context, id string) (*domain.Product, error) {
	product, err := r.store.GetProduct(ctx, id)
	if err != nil {
		return nil, domain.NewInternalError("failed to fetch product", err)
	}
	return product, nil
}

func (r *ProductRepository) CreateProduct(ctx context.Context, product domain.Product) (*domain.Product, error) {
	product = r.ensureTimestamps(product, true)
	created, err := r.store.CreateProduct(ctx, product)
	if err != nil {
		return nil, domain.NewInternalError("failed to save product", err)
	}
	return created, nil
}

func (r *ProductRepository) UpdateProduct(ctx context.Context, id string, product domain.Product) (*domain.Product, error) {
	product = r.ensureTimestamps(product, false)
	updated, err := r.store.UpdateProduct(ctx, id, product)
	if err != nil {
		return nil, domain.NewInternalError("failed to update product", err)
	}
	return updated, nil
}

func (r *ProductRepository) DeleteProduct(ctx context.Context, id string) error {
	deleted, err := r.store.DeleteProduct(ctx, id)
	if err != nil {
		return domain.NewInternalError("failed to delete product", err)
	}
	if !deleted {
		return domain.NewNotFoundError("product not found")
	}
	return nil
}

func (r *ProductRepository) Health(ctx context.Context) error {
	if err := r.store.Health(ctx); err != nil {
		return domain.NewInternalError("store unhealthy", err)
	}
	return nil
}

func (r *ProductRepository) ensureTimestamps(product domain.Product, isCreate bool) domain.Product {
	if r.clock == nil {
		return product
	}
	if product.CreatedAt.IsZero() && isCreate {
		product.CreatedAt = r.clock.Now()
	}
	if product.UpdatedAt.IsZero() {
		if isCreate {
			product.UpdatedAt = product.CreatedAt
		} else {
			product.UpdatedAt = r.clock.Now()
		}
	}
	return product
}

File: node-service/internal/infrastructure/repository/productRepository.js
const DomainError = require('../../domain/errors');

class ProductRepository {
  constructor(store, clock) {
    this.store = store;
    this.clock = clock;
  }

  async list() {
    try {
      return await this.store.listProducts();
    } catch (err) {
      throw DomainError.internal('failed to list products', err);
    }
  }

  async getById(id) {
    try {
      return await this.store.getProduct(id);
    } catch (err) {
      throw DomainError.internal('failed to fetch product', err);
    }
  }

  async create(product) {
    try {
      const now = this.clock?.now();
      if (product.createdAt == null && now) {
        product.createdAt = now;
      }
      if (product.updatedAt == null && now) {
        product.updatedAt = now;
      }
      return await this.store.createProduct(product);
    } catch (err) {
      throw DomainError.internal('failed to create product', err);
    }
  }

  async update(id, product) {
    try {
      if (product.updatedAt == null && this.clock) {
        product.updatedAt = this.clock.now();
      }
      return await this.store.updateProduct(id, product);
    } catch (err) {
      throw DomainError.internal('failed to update product', err);
    }
  }

  async delete(id) {
    try {
      return await this.store.deleteProduct(id);
    } catch (err) {
      throw DomainError.internal('failed to delete product', err);
    }
  }

  async health() {
    await this.store.health();
  }
}

module.exports = ProductRepository;
